cpp/include/cugraph/prims/copy_v_transform_reduce_in_out_nbr.cuh

```diff
@@ -513,23 +516,98 @@ void copy_v_transform_reduce_nbr(raft::handle_t const& handle,
+      for (size_t i = 1; i < graph_view.get_number_of_local_adj_matrix_partitions(); ++i) {
+        assert(graph_view.get_local_adj_matrix_partition_segment_offsets(i));
+      }
+
+      auto& col_comm = handle.get_subcomm(cugraph::partition_2d::key_naming_t().col_name());
+      auto const col_comm_size = col_comm.get_size();
+
+      // memory footprint vs parallelism trade-off
+      // peak memory requirement per loop is
+      // update_major ? V / comm_size * sizeof(T) : 0
+      // and limit memory requirement to (E / comm_size) * sizeof(vertex_t)
+
+      size_t num_streams = std::min(static_cast<size_t>(col_comm_size) * max_segments,
+                                    (handle.get_stream_pool_size() / max_segments) * max_segments);
+      if constexpr (update_major) {
+        size_t value_size{0};
+        if constexpr (is_thrust_tuple_of_arithmetic<T>::value) {
+          auto elem_sizes = compute_thrust_tuple_element_sizes<T>{}();
+          value_size      = std::reduce(elem_sizes.begin(), elem_sizes.end());
+        } else {
+          value_size = sizeof(T);
+        }
+
+        auto avg_vertex_degree = graph_view.get_number_of_vertices() > 0
+                                   ? (static_cast<double>(graph_view.get_number_of_edges()) /
+                                      static_cast<double>(graph_view.get_number_of_vertices()))
+                                   : double{0.0};
+
+        num_streams =
+          std::min(static_cast<size_t>(avg_vertex_degree * (static_cast<double>(sizeof(vertex_t)) /
+                                                            static_cast<double>(value_size))) *
+                     max_segments,
+                   num_streams);
+      }
+
+      if (num_streams >= max_segments) {
+        stream_pool_indices = std::vector<size_t>(num_streams);
+        std::iota((*stream_pool_indices).begin(), (*stream_pool_indices).end(), size_t{0});
+        handle.sync_stream();
+      }
+    }
+  }
+
+  std::vector<decltype(allocate_dataframe_buffer<T>(0, rmm::cuda_stream_view{}))>
+    major_tmp_buffers{};
+  if constexpr (GraphViewType::is_multi_gpu && update_major) {
+    std::vector<size_t> major_tmp_buffer_sizes(
+      graph_view.get_number_of_local_adj_matrix_partitions(), size_t{0});
+    for (size_t i = 0; i < graph_view.get_number_of_local_adj_matrix_partitions(); ++i) {
+      major_tmp_buffer_sizes[i] = GraphViewType::is_adj_matrix_transposed
+                                    ? graph_view.get_number_of_local_adj_matrix_partition_cols(i)
+                                    : graph_view.get_number_of_local_adj_matrix_partition_rows(i);
+    }
+    if (stream_pool_indices) {
+      auto num_concurrent_loops = (*stream_pool_indices).size() / max_segments;
+      major_tmp_buffers.reserve(num_concurrent_loops);
+      for (size_t i = 0; i < num_concurrent_loops; ++i) {
+        size_t max_size{0};
+        for (size_t j = i; j < graph_view.get_number_of_local_adj_matrix_partitions();
```

Comment:
Why not do a `max_element()` here  (lines 581-584)? It's more readable and possibly faster. I understand you have a stride, but you can pass a sequence and calculate the stride inside a lambda comparer, etc.

Reply:
How can I pass a sequence?

Can we do this without using thrust/boost (i.e. without using counting_iterator/transform_iterator).

We can create an additional vector storing a sequence, but then, I am not sure the code will be more readable.

Could you show me the code?

Reply:
You can use counting_iterator/transform_iterator on std::vector<> with thrust::host policy and, say, thrust::reduce() or thrust::maximum_element() but the counter arithmetic can be messy. You're right the resulting code would probably _NOT_ be any more readable.